src/core/SampleBuffer.cpp

```diff
@@ -23,1605 +23,177 @@
+	m_data = result;
+	m_audioFile = audioFile;
+	m_sampleRate = engineRate;
 }
 
-
-void SampleBuffer::normalizeSampleRate(const sample_rate_t srcSR, bool keepSettings)
+QString SampleBuffer::toBase64() const
 {
-	const sample_rate_t oldRate = m_sampleRate;
-	// do samplerate-conversion to our default-samplerate
-	if (srcSR != audioEngineSampleRate())
-	{
-		SampleBuffer * resampled = resample(srcSR, audioEngineSampleRate());
-
-		m_sampleRate = audioEngineSampleRate();
-		MM_FREE(m_data);
-		m_frames = resampled->frames();
-		m_data = MM_ALLOC<sampleFrame>( m_frames);
-		memcpy(m_data, resampled->data(), m_frames * sizeof(sampleFrame));
-		delete resampled;
-	}
-
-	if (keepSettings == false)
-	{
-		// update frame-variables
-		m_loopStartFrame = m_startFrame = 0;
-		m_loopEndFrame = m_endFrame = m_frames;
-	}
-	else if (oldRate != audioEngineSampleRate())
-	{
-		auto oldRateToNewRateRatio = static_cast<float>(audioEngineSampleRate()) / oldRate;
-
-		m_startFrame = std::clamp(f_cnt_t(m_startFrame * oldRateToNewRateRatio), 0, m_frames);
-		m_endFrame = std::clamp(f_cnt_t(m_endFrame * oldRateToNewRateRatio), m_startFrame, m_frames);
-		m_loopStartFrame = std::clamp(f_cnt_t(m_loopStartFrame * oldRateToNewRateRatio), 0, m_frames);
-		m_loopEndFrame = std::clamp(f_cnt_t(m_loopEndFrame * oldRateToNewRateRatio), m_loopStartFrame, m_frames);
-		m_sampleRate = audioEngineSampleRate();
-	}
+	// TODO: Replace with non-Qt equivalent
+	const auto data = reinterpret_cast<const char*>(m_data.data());
+	const auto size = static_cast<int>(m_data.size() * sizeof(sampleFrame));
+	const auto byteArray = QByteArray{data, size};
+	return byteArray.toBase64();
 }
 
-
-
-
-f_cnt_t SampleBuffer::decodeSampleSF(
-	QString fileName,
-	sample_t * & buf,
-	ch_cnt_t & channels,
-	sample_rate_t & samplerate
-)
+auto SampleBuffer::audioFile() const -> QString
 {
-	SNDFILE * sndFile;
-	SF_INFO sfInfo;
-	sfInfo.format = 0;
-	f_cnt_t frames = 0;
-	sf_count_t sfFramesRead;
-
-
-	// Use QFile to handle unicode file names on Windows
-	QFile f(fileName);
-	if (f.open(QIODevice::ReadOnly) && (sndFile = sf_open_fd(f.handle(), SFM_READ, &sfInfo, false)))
-	{
-		frames = sfInfo.frames;
-
-		buf = new sample_t[sfInfo.channels * frames];
-		sfFramesRead = sf_read_float(sndFile, buf, sfInfo.channels * frames);
-
-		if (sfFramesRead < sfInfo.channels * frames)
-		{
-#ifdef DEBUG_LMMS
-			qDebug("SampleBuffer::decodeSampleSF(): could not read"
-				" sample %s: %s", fileName, sf_strerror(nullptr));
-#endif
-		}
-		channels = sfInfo.channels;
-		samplerate = sfInfo.samplerate;
-
-		sf_close(sndFile);
-	}
-	else
-	{
-#ifdef DEBUG_LMMS
-		qDebug("SampleBuffer::decodeSampleSF(): could not load "
-				"sample %s: %s", fileName, sf_strerror(nullptr));
-#endif
-	}
-	f.close();
-
-	//write down either directly or convert i->f depending on file type
-
-	if (frames > 0 && buf != nullptr)
-	{
-		directFloatWrite(buf, frames, channels);
-	}
-
-	return frames;
+	return m_audioFile.value_or("");
```

Comment:
It would, though it was simpler to just use an empty string to represent no audio file. I do think that ``std::optional`` reads better, as it explicitly shows that this ``SampleBuffer`` may or may not have an audio file. I'll change it soon.

Reply:
That would also mean I would have to use ``value_or`` everywhere else where the callers are okay with an empty string meaning no audio file. I am not sure what's the best approach. On one side, returning ``std::optional`` reads better, while returning an empty ``QString`` for no file means we don't have to put ``value_or("")`` everywhere.

Reply:
I think most people using the method would understand that an empty string means no audio file, so I would just go with `QString`. It's more convenient too. And the `m_audioFile` member variable could also be a `QString` for the same reason + for consistency.